File: sdks/go/cache_key.go
package lcp

import (
	"errors"
	"strings"

	b57 "github.com/aco/f57"
)

type HashFn func([]byte) string

type CacheKeyInput struct {
	Namespace     string
	OperationID   string
	Payload       any
	SchemaVersion string
	SpecChecksum  string
	UserScope     string
}

var traceFields = map[string]struct{}{
	"trace_id":   {},
	"action_id":  {},
	"request_id": {},
}

var ErrHashFnRequired = errors.New("hashFn is required (use H57HashFn)")

func stripTraceFields(value any) any {
	switch v := normalizeTraversalValue(value).(type) {
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = stripTraceFields(item)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(v))
		for key, item := range v {
			if _, ok := traceFields[key]; ok {
				continue
			}
			out[key] = stripTraceFields(item)
		}
		return out
	default:
		return value
	}
}

func BuildCacheKeyMaterial(input CacheKeyInput) (string, error) {
	payloadJSON, err := CanonicalJSONStringify(stripTraceFields(input.Payload))
	if err != nil {
		return "", err
	}
	parts := []string{
		input.Namespace,
		input.OperationID,
		payloadJSON,
		input.SchemaVersion,
		input.SpecChecksum,
		input.UserScope,
	}
	return strings.Join(parts, "|"), nil
}

func ComputeCacheKey(input CacheKeyInput, hashFn HashFn) (string, error) {
	material, err := BuildCacheKeyMaterial(input)
	if err != nil {
		return "", err
	}
	if hashFn == nil {
		return "", ErrHashFnRequired
	}
	return hashFn([]byte(material)), nil
}

func DefaultHashFn(bytes []byte) string {
	result, err := b57.H57Hash(bytes, b57.H57HashAuto)
	if err != nil {
		panic("H57Hash: " + err.Error())
	}
	return result
}

// H57HashFn is the canonical H57 hash hook: BLAKE3 → B57 encoding.
// Uses github.com/aco/f57 mapped to the F57 reference implementation.
func H57HashFn(bytes []byte) string {
	return DefaultHashFn(bytes)
}

File: sdks/flutter/lib/src/key/cache_key.dart
import 'dart:convert';

import 'canonical_json.dart';

typedef HashFn = String Function(List<int> bytes);

class CacheKeyInput {
  const CacheKeyInput({
    required this.namespace,
    required this.operationId,
    required this.payload,
    required this.schemaVersion,
    required this.specChecksum,
    required this.userScope,
  });

  final String namespace;
  final String operationId;
  final Object? payload;
  final String schemaVersion;
  final String specChecksum;
  final String userScope;
}

const _traceFields = <String>{'trace_id', 'action_id', 'request_id'};

Object? _stripTraceFields(Object? value) {
  if (value is List) {
    return value.map(_stripTraceFields).toList(growable: false);
  }
  if (value is Map) {
    final output = <String, Object?>{};
    value.forEach((key, item) {
      final normalizedKey = key.toString();
      if (_traceFields.contains(normalizedKey)) return;
      output[normalizedKey] = _stripTraceFields(item);
    });
    return output;
  }
  return value;
}

String buildCacheKeyMaterial(CacheKeyInput input) {
  final payloadJson = canonicalJsonStringify(_stripTraceFields(input.payload));
  return [
    input.namespace,
    input.operationId,
    payloadJson,
    input.schemaVersion,
    input.specChecksum,
    input.userScope,
  ].join('|');
}

String computeCacheKey(CacheKeyInput input, HashFn hashFn) {
  final material = buildCacheKeyMaterial(input);
  return hashFn(utf8.encode(material));
}
